gcs/src/main/scala/monix/connect/gcp/storage/components/GcsUploader.scala

```diff
@@ -46,44 +45,61 @@ private[storage] final class GcsUploader(
   override def createSubscriber(
     cb: Callback[Throwable, Unit],
     s: Scheduler): (Subscriber[Array[Byte]], AssignableCancelable) = {
-    val sub = new Subscriber[Array[Byte]] {
+    val sub = new Subscriber.Sync[Array[Byte]] {
       self =>
       override implicit def scheduler: Scheduler = s
 
       val writer: WriteChannel = storage.writer(blobInfo, options: _*)
       writer.setChunkSize(chunkSize)
+      private[this] var isDone = false
 
-      override def onNext(chunk: Array[Byte]): Future[Ack] = {
-        Task {
+      override def onNext(chunk: Array[Byte]): Ack = {
+        try {
+          if (chunk.isEmpty) {
+            if (!isDone) { onComplete() }
```

Comment:
the flag is already checked in `onComplete`

Reply:
should we even call onComplete here? or better to leave that responsability to the observable?

Reply:
No, it's fine - we call our own `onComplete` and then return `Stop` so even if `Observable` still sends `onComplete`, we should ignore it (which we do with isDone flag).

We need to call `onComplete` on our own because if `Observable` notices `Stop`, it most likely won't submit anything, it will shut down due to our "order", the only exception is if it calls:

```
out.onNext(elem)
out.onComplete
```

And it would only happen if we (downstream) stop at the same time as upstream

Reply:
thanks for the detailed explanation :-)